service-csw/src/main/java/org/oskari/csw/request/GetRecords.java

```diff
@@ -0,0 +1,177 @@
+            xsw.writeEndElement(); // ElementSetName
+
+
+            xsw.writeStartElement(CSW_URI, "Constraint");
+            xsw.writeAttribute("version", CONSTRAINT_VERSION);
+        } catch (XMLStreamException e) {
+            throw new ServiceRuntimeException("Couldn't create GetRecords request", e);
+        }
+    }
+
+    /**
+     * Returns the generated XML.
+                ...includes filters... and end document
+             </csw:Constraint>
+         </csw:Query>
+     </csw:GetRecords>
+     * @return
+     */
+    private static void endDocument(XMLStreamWriter xsw) {
+        try {
+            xsw.writeEndElement(); // Constraint
+            xsw.writeEndElement(); // Query
+            xsw.writeEndElement(); // GetRecords
+            xsw.writeEndDocument();
+            xsw.close();
+        } catch (XMLStreamException e) {
+            throw new ServiceRuntimeException("Couldn't create GetRecords request", e);
+        }
+    }
+
+    /**
+     * Serialize the filter to get a String that can be injected into the stream.
+     * @param filter generated with Geotools
+     * @return String that can be injected to XML document
+     */
+    private static String getFilterAsString(Filter filter) {
+        if (filter == null) {
+            throw new ServiceRuntimeException("Tried to serialize null-filter");
+        }
+        try {
+            Encoder encoder = new Encoder(new OGCConfiguration());
+            encoder.setOmitXMLDeclaration(true);
+            return encoder.encodeAsString(filter, org.geotools.filter.v1_0.OGC.Filter);
+        } catch (IOException e) {
+            throw new ServiceRuntimeException("Couldn't serialize filter to string", e);
+        }
+    }
+
+    /**
+     * Injects any string in the middle of the XML document.
+     *
+     * Unsafe because injecting invalid XML-fragment as string makes the whole document invalid
+     * and you can do it with this method.
+     * @param xsw
+     * @param stream
+     * @param injectUnsafe XML fragment to inject in the middle of the stream.
+     */
+    private static void writeRawXMLUnsafe(XMLStreamWriter xsw, OutputStream stream, String injectUnsafe) {
+        try {
+            // Following line is very important!!
+            // without it unescaped data will appear inside the previously opened tag.
+            xsw.writeCharacters("");
+            xsw.flush();
+
+            OutputStreamWriter osw = new OutputStreamWriter(stream);
+            osw.write(injectUnsafe);
+            osw.flush();
+        } catch (XMLStreamException | IOException e) {
+            throw new ServiceRuntimeException("Couldn't write raw XML to GetRecords request", e);
+        }
+    }
+
+    /**
+     * Serialize stream to get a String
+     * @param stream
+     * @return
+     */
+    private static String getAsString(ByteArrayOutputStream stream) {
+        return new String (stream.toByteArray(), Charset.forName("UTF-8"));
```

Comment:
See `ByteArrayOutputStream#toString(StandardCharsets.UTF_8.name())`

Reply:
Better yet if we want to return a `String` from `createRequest(Filter)` we can use `StringWriter` instead of `ByteArrayOutputStream` here and avoid encoding and decoding chars to UTF-8 bytes and back.

Reply:
Or we can change the function header to `ByteArrayOutputStream createRequest(Filter, Charset)` and `MetadataCatalogueChannelSearchService` could then use `IOHelper.post(String url, String contentType,         ByteArrayOutputStream baos)`. This would be the most efficient way of doing it.
If in unit tests we do want Strings then we can easily create them there from the ByteArrayOutputStream.

Reply:
While this certainly isn't performance critical code I think we should still strive for (the most) efficient implementation throughout the codebase while maintaining high readability.
In my opinion passing around a ByteArrayOutputStream instead of a String would not make this code any less readable.